### DISCUSSION TITLE
Example 15: Real PushButtons --> changing PWM step and time

### DISCUSSION THREAD

**Discussion by leotordo** [HomeSpan/HomeSpan#529] [Q&A]
Hello,
 I'm writing code for my application starting from "Example 15: Real PushButtons".
With DEV_DimmableLED the long press of raise/lower button the led dimming changes up and down with 10% steps in about 5 seconds.

I need these changes
1) reduce long press step to 1%
2) set full dimming ramp in about 3 seconds.

Where I've to do these changes?
Thanks
Stefano

**Comment by HomeSpan** [comment]
The logic for what each button does, increments, etc. are all coded within the `void button()` method of the DEV_DimmableLED class.  You may change this to any combination of logic desired.  Here, for example, is the logic that says a LONG press of the raise button increases brightness by 10, else use an increment of 1:

```C++

    if(pin==raisePin){                                                   
      if(pressType==SpanButton::DOUBLE){            // if a DOUBLE press of the raise button...
        power->setVal(1);                           // ...turn on power
        level->setVal(100);                         // ...and set brightness to the max level
      } else {
      
      newLevel=level->getVal()+(pressType==SpanButton::LONG?10:1);   // get current brightness level and increase by either 10% (LONG press) or 1% (SINGLE press)
      if(newLevel>100)                                               // don't allow new level to exceed maximium of 100%
        newLevel=100;
      level->setVal(newLevel);                                       // set the value of the brightness Characteristic to this new level
      }
```

The speed at which a LONG press is recognized is determined by the second parameter of the SpanButton constructor.  Default is 2000ms, though it is changed in the code to 500ms in the following line:

```C++
    new SpanButton(raisePin,500);                   // NEW! create new SpanButton to increase brightness using pushbutton on pin number "raisePin"
```

See the [API Reference](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Reference.md#spanbuttonint-pin-uint16_t-longtime-uint16_t-singletime-uint16_t-doubletime-boolean-triggertypeint) for details on SpanButton.

**Reply by leotordo** [reply]
Hello. Thanks for suggestions.

On API Reference I found informations, now my application works quite good.
There a little problem. The best combination of ramp speed, smoothing and button times is
300 for SHORT
100 for LONG
Unfortunately it doesn't work like reported in API Reference

_If singleTime>longTime, only Long Press triggers can occur_

is there a workaround? I need long time for SHORT press and speed up the dimming ramp without increasing steps

Thank you very much
Stefano

**Comment by HomeSpan** [comment]
I'm not sure I understand what timings you are trying to do, or what's not working according to the API.   You can't operationally have a short-press time that is longer than the long-press time.  If a short press is 300 milliseconds, the long press will be already triggered after 100 milliseconds and the short press will never be activated.

It seems that you may be trying to use the LONG press as an **automatic** fader to go from 0-100% or 100-0% in steps of 1%.  If so, that's not what the code is designed to do - the button is meant to be held, which necessarily means it will take a while for each step to be processed.  That's by design.

If you are trying to create an automatic fading mechanism, you'll need to considerably change the logic so that a LONG press (or another button) triggers some code you write to perform the automatic fading after the button is pressed just one time.  In writing this code you need to ramp the brightness from 0-100% (or 100-0%) BUT this cannot be done in a simple loop --- if your ramp time is more than 1 second or so, HomeKit will probably complain that it has not received a response from HomeSpan.  You'll need to create a non-blocking loop that uses `millis()` or something like that to change the brightness according to a schedule without blocking the code.  I don't have any examples of that at present, though will consider that as a future expansion of the PWM library.

Also, for automated fading you should not send `setVal()` at every increment.  That can overload HomeKit.  Instead, perform the full ramp to 0% or 100% and than send a single `setVal()` at the end.

**Reply by leotordo** [reply]
> I'm not sure I understand what timings you are trying to do, or what's not working according to the API. You can't operationally have a short-press time that is longer than the long-press time. If a short press is 300 milliseconds, the long press will be already triggered after 100 milliseconds and the short press will never be activated.
It seems that you may be trying to use the LONG press as an automatic fader to go from 0-100% or 100-0% in steps of 1%. If so, that's not what the code is designed to do - the button is meant to be held, which necessarily means it will take a while for each step to be processed. That's by design.

Yes, you're right. I'm trying to use LONG for a fader 0-100 and back

>If you are trying to create an automatic fading mechanism, you'll need to considerably change the logic so that a LONG press (or another button) triggers some code you write to perform the automatic fading after the button is pressed just one time. In writing this code you need to ramp the brightness from 0-100% (or 100-0%) BUT this cannot be done in a simple loop --- if your ramp time is more than 1 second or so, HomeKit will probably complain that it has not received a response from HomeSpan. You'll need to create a non-blocking loop that uses millis() or something like that to change the brightness according to a schedule without blocking the code. I don't have any examples of that at present, though will consider that as a future expansion of the PWM library.

ok, I think I understood

> Also, for automated fading you should not send setVal() at every increment. That can overload HomeKit. Instead, perform the full ramp to 0% or 100% and than send a single setVal() at the end.

Thanks for this, I'm not skilled on HomeKit applications.

I'll try to write new code for my specifications.

I'll update this thread when I've some news.

Thanks
Stefano
